src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs

```diff
@@ -1361,29 +1362,43 @@ private void DecodeCustomAttributeParameterTypeOrThrow(ref BlobReader sigReader,
 
                 case SignatureTypeCode.TypeHandle:
                     // The type of the parameter can either be an enum type or System.Type.
-                    bool isNoPiaLocalType;
                     type = GetSymbolForTypeHandleOrThrow(sigReader.ReadTypeHandle(), out isNoPiaLocalType, allowTypeSpec: true, requireShortForm: true);
+                    break;
 
-                    var underlyingEnumType = GetEnumUnderlyingType(type);
-
-                    // Spec: If the parameter kind is an enum -- simply store the value of the enum's underlying integer type.
-                    if ((object)underlyingEnumType != null)
+                case SignatureTypeCode.GenericTypeParameter:
+                    if (!sigReader.TryReadCompressedInteger(out int paramPosition))
                     {
-                        Debug.Assert(!isNoPiaLocalType);
-
-                        // GetEnumUnderlyingType always returns a valid enum underlying type
-                        typeCode = underlyingEnumType.SpecialType.ToSerializationType();
-                        return;
+                        throw new UnsupportedSignatureContent();
                     }
 
-                    if ((object)type == SystemTypeSymbol)
-                    {
-                        Debug.Assert(!isNoPiaLocalType);
-                        typeCode = SerializationTypeCode.Type;
-                        return;
-                    }
+                    type = GetGenericTypeArgumentSymbol(paramPosition);
```

Comment:
>type = GetGenericTypeArgumentSymbol(paramPosition); [](http://example.com/codeflow?start=20&length=51)

This looks suspicious. Why are we using generic type argument instead of parameter?

Reply:
I'll stop reviewing until I understand what is going on here. It looks like this part is fundamental for the fix. Let's sync up offline.

Reply:
For example, what is going to happen if this is a reference to a generic type parameter of the enclosing type instead?

Reply:
> Why are we using generic type argument instead of parameter?

If we have attribute like `class A<T>(T arg) : System.Attribute` used like `[A<B<int>.E>(B<int>.C)]`, the attribute constructor argument has type `T` in metadata (`GenericTypeParameter`) which needs to be decoded as `B<int>.E` (the corresponding argument).

> what is going to happen if this is a reference to a generic type parameter of the enclosing type instead?

This is tested in `Attribute_GenericParameter_NestedClass`. The `AdjustGenericTypeParameterPosition` method handles this - going through enclosing types until the corresponding type argument is found. Note that this logic already existed in `GetGenericTypeParamSymbol`, I have just extracted it into a reusable method.

This is tested in `Attribute_GenericParameter_NestedClass` - for example there the full attribute type (and hence the `_typeContextOpt`) is `A1<B<System.Int32>.E>.A2` and `paramPosition` is 0.